airflow/contrib/operators/kubernetes_pod_operator.py

```diff
@@ -112,55 +113,60 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
     """
     template_fields = ('cmds', 'arguments', 'env_vars', 'config_file')
 
+    @staticmethod
+    def create_labels_for_pod(context):
+        """
+        Generate labels for the pod s.t. we can track it in case of Operator crash
+
+        :param context:
+        :return:
+        """
+        labels = {
+            'dag_id': context['dag'].dag_id,
+            'task_id': context['task'].task_id,
+            'exec_date': context['ts'],
+            'try_number': context['ti'].try_number,
```

Comment:
I think we need to consider the desired behaviour here.

What could happen now is that we create duplicate pods in the case of a previous try still running e.g. in the case of task timeout & no kill or failed kill.  In that case we introduce a new scenario of duplicate pods which is what we're trying to solve.  Certainly for my jobs this would cause issues as we're not using try_number to create output locations for example.

One option is to check for the previous try pod first and kill it if it exists before running the next one.

Reply:
The counter argument to this: if we _don't_ include try_number here then it could "reattach" to a previous running pod when Airflow thinks it has retired. That seems like an odd behaviour.

This is all edge cases anyway, it's unlikely that we would be in this situation.

What we could do is search by dag_id/task_id/exec_date (i.e. don't search by try_number) and then if we find a running pod with something other than the expected try_number we kill it? A little bit worried we might kill the wrong pod somehow there, but it _should_ be safe?

Reply:
I agree with all that you've said, I think that is a good solution and should be safe.

Reply:
I'm now second guessing myself on this.

If we do add this check on try_number in I can't think of a case when we would re-attach to a running pod. (because if this operator fails/dies/zombies for it to run again it would be the next try_number.)

Reply:
I should've realised that.  We do indeed monitor a previous try number on purpose in my original changes.

With the try number changes the main duplicate pod issue will be fixed still as we force kill it before starting a new one.  But I took the view that, although perhaps confusing in terms of try number, it is far more efficient to continue monitoring the previous try in the case of worker failure.  AFAIK worker failure/restart is the only situation in which pods remain running and are not killed.  If you fail a task via the UI it triggers a pod kill for example.

Unless you have another solution that can check & re-use the previous try number then I think we're better off removing the try_number changes?

Reply:
Hey @ashb @dimberman how do feel about removing the try_number changes and allowing the re-attach to a previous pod?  This is the logic I've been running for a while now.  The situation I can think of where you'd not want this is if the pod hangs for some reason (sub process failure or similar) and then a kill on timeout fails.  Quite unlikely... but to consider.

Reply:
@danccooper I've definitely run into that scenario